pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java

```diff
@@ -218,30 +219,33 @@ public static SegmentSelectionResult processValidDocIdsMetadata(Map<String, Stri
             String.valueOf(DEFAULT_INVALID_RECORDS_THRESHOLD_COUNT)));
     List<Pair<SegmentZKMetadata, Long>> segmentsForCompaction = new ArrayList<>();
     List<String> segmentsForDeletion = new ArrayList<>();
-    for (ValidDocIdsMetadataInfo validDocIdsMetadata : validDocIdsMetadataInfoList) {
-      long totalInvalidDocs = validDocIdsMetadata.getTotalInvalidDocs();
-      String segmentName = validDocIdsMetadata.getSegmentName();
-
-      // Skip segments if the crc from zk metadata and server does not match. They may be being reloaded.
-      SegmentZKMetadata segment = completedSegmentsMap.get(segmentName);
-      if (segment == null) {
-        LOGGER.warn("Segment {} is not found in the completed segments list, skipping it for compaction", segmentName);
-        continue;
-      }
+    for (String segmentName: validDocIdsMetadataInfoMap.keySet()) {
+      for (ValidDocIdsMetadataInfo validDocIdsMetadata : validDocIdsMetadataInfoMap.get(segmentName)) {
+        long totalInvalidDocs = validDocIdsMetadata.getTotalInvalidDocs();
+
+        // Skip segments if the crc from zk metadata and server does not match. They may be being reloaded.
+        SegmentZKMetadata segment = completedSegmentsMap.get(segmentName);
```

Comment:
this check can be out of the inner for-loop, and continue to next segment `if(!completedSegmentsMap.contains(segmentName))`

Reply:
Nice catch!